ngraph/test/type_prop/reshape.cpp

```diff
@@ -158,3 +140,132 @@ TEST(type_prop, reshape_partial_rank_static_dynamic_but_zero_ok)
+    auto r = make_shared<op::v1::Reshape>(
+        param, op::Constant::create(element::u64, {2}, std::vector<int64_t>{0, -1}), true);
+    ASSERT_EQ(r->get_element_type(), element::f32);
+    ASSERT_EQ(r->get_output_partial_shape(0), (PartialShape{Dimension::dynamic(), 1}));
+}
+
+TEST(type_prop, reshape_partial_rank_dynamic_special_zero)
+{
+    auto param = make_shared<op::Parameter>(element::f32, PartialShape::dynamic());
+    auto r = make_shared<op::v1::Reshape>(
+        param, op::Constant::create(element::u64, {4}, std::vector<int64_t>{3, 1, 0, 2}), true);
+    ASSERT_EQ(r->get_element_type(), element::f32);
+    ASSERT_EQ(r->get_output_partial_shape(0), (PartialShape{3, 1, Dimension::dynamic(), 2}));
+}
+
+TEST(type_prop, reshape_partial_rank_dynamic_special_neg)
+{
+    auto param = make_shared<op::Parameter>(element::f32, PartialShape::dynamic());
+    auto r = make_shared<op::v1::Reshape>(
+        param, op::Constant::create(element::u64, {4}, std::vector<int64_t>{3, -1, 0, 2}), true);
+    ASSERT_EQ(r->get_element_type(), element::f32);
+    ASSERT_EQ(r->get_output_partial_shape(0),
+              (PartialShape{3, Dimension::dynamic(), Dimension::dynamic(), 2}));
+}
+
+TEST(type_prop, reshape_deduce_special_zero_shape_zero_zero_one_neg_dynamic_with_interval)
+{
+    auto param = make_shared<op::Parameter>(element::f32, PartialShape{2, Dimension(1, 3), 3});
+    auto r = make_shared<op::v1::Reshape>(
+        param, op::Constant::create(element::u64, {4}, std::vector<int64_t>{0, 0, 1, -1}), true);
+    ASSERT_EQ(r->get_element_type(), element::f32);
+    ASSERT_EQ(r->get_output_partial_shape(0), (PartialShape{2, Dimension(1, 3), 1, 3}));
+}
+
+TEST(type_prop, reshape_deduce_special_zero_shape_zero_zero_one_neg_double_dynamic_with_interval)
+{
+    auto param = make_shared<op::Parameter>(element::f32,
+                                            PartialShape{2, Dimension(1, 3), Dimension::dynamic()});
+    auto r = make_shared<op::v1::Reshape>(
+        param, op::Constant::create(element::u64, {4}, std::vector<int64_t>{0, 0, 1, -1}), true);
+    ASSERT_EQ(r->get_element_type(), element::f32);
+    ASSERT_EQ(r->get_output_partial_shape(0),
+              (PartialShape{2, Dimension(1, 3), 1, Dimension::dynamic()}));
+}
+
+TEST(type_prop, reshape_deduce_special_zero_shape_zero_neg_dynamic_with_interval)
+{
+    auto param = make_shared<op::Parameter>(element::f32, PartialShape{2, Dimension(1, 3)});
+    auto r = make_shared<op::v1::Reshape>(
+        param, op::Constant::create(element::u64, {2}, std::vector<int64_t>{0, -1}), true);
+    ASSERT_EQ(r->get_element_type(), element::f32);
+    ASSERT_EQ(r->get_output_partial_shape(0), (PartialShape{2, Dimension::dynamic()}));
+}
+
+TEST(type_prop, reshape_deduce_special_zero_shape_neg_zero_dynamic_with_interval)
+{
+    auto param = make_shared<op::Parameter>(element::f32, PartialShape{2, Dimension(1, 3)});
```

Comment:
Please add tests with interval dims and -1 in target shape:
```suggestion
   TEST(type_prop, reshape_deduce_special_zero_shape_neg_zero_dynamic_with_interval)
{
    auto param = make_shared<op::Parameter>(element::f32, PartialShape{Dimension(1, 3), 2});
    auto r = make_shared<op::v1::Reshape>(
    param, op::Constant::create(element::u64, {2}, std::vector<int64_t>{-1, 0}), true);
    ASSERT_EQ(r->get_element_type(), element::f32);
    ASSERT_EQ(r->get_output_partial_shape(0), (PartialShape{Dimension(1, 3), 2}));
}
```